### PR TITLE
Redundant copying of weights and biases for convolution layer

### PR DESCRIPTION
**Pull request by vladislav-volkov**
### Details:
 - Redundant copying of weights and biases removed for convolution layer.
 - Some legacy code removed from MKLDNN convolution layer

### Tickets:
 - CVS-46997

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h [hunk 1]
```diff
@@ -21,19 +21,18 @@ class MKLDNNInputNode : public MKLDNNNode {
...
+    MKLDNNMemoryPtr getMemoryPtr() const;
 
-    const InferenceEngine::Blob::CPtr getConstBlob() const {
-        return constBlob;
-    }
+private:
+    void cloneBlobIfRequired(const MKLDNNDims& dims, const InferenceEngine::Precision& prec);
 
 private:
+    std::shared_ptr<ngraph::Node> origLayer;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h:32]: Seems like unused.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h [hunk 2]
```diff
@@ -21,19 +21,18 @@ class MKLDNNInputNode : public MKLDNNNode {
     void createPrimitive() override;
     bool created() const override;
 
-    void execute(mkldnn::stream strm) override;
-    void withMeanImage() {
-        isMeanImage = true;
-    }
+    void withMeanImage();
+    const InferenceEngine::Blob::CPtr getConstBlob() const;
+    MKLDNNMemoryPtr getMemoryPtr() const;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h:26]: Shouldn't it be const MKLDNNMemoryPtr?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp [hunk 3]
```diff
@@ -48,10 +301,92 @@ MKLDNNInputNode::MKLDNNInputNode(const std::shared_ptr<ngraph::Node>& op, const
...
+    auto blobKey = [this] () {
+        char ptr[32];
+        snprintf(ptr, sizeof ptr, "%p", constBlob->cbuffer().as<const void*>());
+        return getName()
+                + "_" + std::to_string(constBlob->byteSize())
+                + "_" + ptr;
+    };
+
+    const void *data = constBlob->buffer();
+    (void)data;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:381]: Why do we need this line?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp [hunk 4]
```diff
@@ -68,6 +403,18 @@ MKLDNNInputNode::MKLDNNInputNode(const InferenceEngine::SizeVector &dims, const
...
 
+void MKLDNNInputNode::withMeanImage() {
+    isMeanImage = true;
+}
+
+const InferenceEngine::Blob::CPtr MKLDNNInputNode::getConstBlob() const {
+    return constBlob;
+}
+
+MKLDNNMemoryPtr MKLDNNInputNode::getMemoryPtr() const {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:414]: Having both `getMemoryPtr` and `getConstBlob` looks like API duplication. Moreover using getConstBlob() might lead to denormals spread throughout the plugin. Can we consider removing `getConstBlob`?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp [hunk 5]
```diff
@@ -3,27 +3,280 @@
...
+        mov(reg_mask_addr, (size_t)mask_data);
+        vmovdqu(mask, yword[reg_mask_addr]);
+
+        // Get arguments addresses
+        mov(reg_src, ptr[param1 + offsetof(args_t, src)]);
+        lea(reg_dst, ptr[param1 + offsetof(args_t, hasSubnormals)]);
+        mov(reg_sz, ptr[param1 + offsetof(args_t, count)]);
+
+        // Main loop
+        xor_(rsi, rsi);
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:134]: does it make sense to get rsi meaningful name as for other registers?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp [hunk 6]
```diff
@@ -3,27 +3,280 @@
...
+            mov(r15d, dword[src + idx * sizeof(float)]);
+            mov(dword[dst + idx * sizeof(float)], r15d);
+        });
+
+        pop(r15);
+        pop(rsi);
+    }
+};
+
+struct jit_has_subnormals_avx2 : public jit_has_subnormals {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:90]: Why you need separate function for sse? They are almost the same. Could you please handle copypaste by using template? We have a lot of good examples in plugin which use this approach.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h (hunk 1 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h:32]: It is necessary to hold a reference to the original buffer inside ngraph::Constant.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h (hunk 2 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h:26]: I think here it isn't mandatory. The const MKLDNNMemoryPtr is just prevent assign to another share pointer.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp (hunk 3 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:381]: removed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp (hunk 4 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:414]: ok, i'll remove it

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp (hunk 6 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp (hunk 5 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h [hunk 7]
```diff
@@ -21,19 +22,16 @@ class MKLDNNInputNode : public MKLDNNNode {
...
-    void withMeanImage() {
-        isMeanImage = true;
-    }
+    void withMeanImage();
+    MKLDNNMemoryCPtr getMemoryPtr() const;
 
-    const InferenceEngine::Blob::CPtr getConstBlob() const {
-        return constBlob;
-    }
+private:
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h:28]: minor: two private sections

**Review by dmitry-gorokhov — APPROVED** [review]

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h (hunk 7 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.h:28]: I use private sections to separate function declarations and class members. (:

**Review by ilyachur — APPROVED** [review]

**Pull request merged**